### PR TITLE
Pass more information in user defined parse error

### PR DESCRIPTION
**Pull request by palaviv**
Before the message was:
```py
Traceback (most recent call last):
SyntaxError: 'Got unsupported token(s) at line 0'
```
And now it is:
```py
Traceback (most recent call last):
SyntaxError: 'Got unexpected token \t at line 4 column 25'
```

### CODE REVIEW

File: compiler/src/error.rs [hunk 1]
```diff
@@ -56,7 +56,10 @@ impl fmt::Display for CompileError {
         }?;
 
         // Print line number:
-        write!(f, " at line {:?}", self.location.row())
+        match &self.error {
+            CompileErrorType::Parse(..) => Ok(()),
```
Comment (windelbouwman) [inline compiler/src/error.rs]: Hmm. This exception for the parse error is not ideal, but we can improve on this later.

File: parser/src/error.rs [hunk 2]
```diff
@@ -24,7 +24,7 @@ pub enum ParseError {
     /// Parser encountered an unexpected token
     UnrecognizedToken(TokSpan, Vec<String>),
     /// Maps to `User` type from `lalrpop-util`
-    Other,
+    Other(LexicalError),
```
Comment (windelbouwman) [inline parser/src/error.rs]: It might be good to rename this error variant into `Lexical(LexicalError)`.

File: parser/src/error.rs [hunk 3]
```diff
@@ -34,8 +34,7 @@ impl From<InnerError<Location, Tok, LexicalError>> for ParseError {
             // TODO: Are there cases where this isn't an EOF?
             InnerError::InvalidToken { location } => ParseError::EOF(Some(location)),
             InnerError::ExtraToken { token } => ParseError::ExtraToken(token),
-            // Inner field is a unit-like enum `LexicalError::StringError` with no useful info
-            InnerError::User { .. } => ParseError::Other,
+            InnerError::User { error } => ParseError::Other(error),
```
Comment (windelbouwman) [inline parser/src/error.rs]: This is very good to propagate the lexer errors!

File: parser/src/lexer.rs [hunk 4]
```diff
@@ -74,6 +75,26 @@ pub enum LexicalErrorType {
...
+        match self.error {
+            LexicalErrorType::StringError => write!(f, "Got unexpected string"),
+            LexicalErrorType::UnicodeError => write!(f, "Got unexpected unicode"),
+            LexicalErrorType::NestingError => write!(f, "Got unexpected nesting"),
+            LexicalErrorType::UnrecognizedToken { tok } => {
+                write!(f, "Got unexpected token {}", tok)
+            }
+            LexicalErrorType::OtherError(ref msg) => write!(f, "{}", msg),
+        }?;
+        write!(
```
Comment (windelbouwman) [inline parser/src/lexer.rs]: I would prefer to add the line and column info later on, so in the `Display` method of the `CompilerError` struct.

File: compiler/src/error.rs (hunk 1 above)
Comment (windelbouwman) [reply compiler/src/error.rs]: Please implement the todo at this line: https://github.com/RustPython/RustPython/blob/ef1d5433df31a3c4ba7e58250b75de85ca76eaa1/compiler/src/error.rs#L17

Then we have proper locations on all compiler errors (where possible).

Comment (palaviv) [reply compiler/src/error.rs]: Done

File: parser/src/error.rs (hunk 2 above)
Comment (palaviv) [reply parser/src/error.rs]: Done

File: parser/src/lexer.rs (hunk 4 above)
Comment (palaviv) [reply parser/src/lexer.rs]: Done

**Comment by windelbouwman** [comment]
This change looks very good now. The only thing left is to take a look at the wasm build.

**Comment by palaviv** [comment]
I am trying to fix the wasm but the error currently look like:
```py
SyntaxError: Error parsing Python code: Got unexpected token: DoubleStar at Location { row: 1, column: 8 } at line 0
```
Not sure why the `fmt::Display` I defined is not working.

File: wasm/lib/src/vm_class.rs [hunk 5]
```diff
@@ -281,30 +281,13 @@ impl WASMVirtualMachine {
...
-                            let _ =
-                                Reflect::set(&js_err, &"endrow".into(), &(loc.row() as u32).into());
-                            let _ = Reflect::set(
-                                &js_err,
-                                &"endcol".into(),
-                                &(loc.column() as u32).into(),
-                            );
-                        }
-                    }
+                    let _ =
```
Comment (windelbouwman) [inline wasm/lib/src/vm_class.rs]: Well, adding the `location` attribute clearly was beneficial!

**Pull request merged**